internal/fxclock/clock_test.go

```diff
@@ -22,44 +22,150 @@ package fxclock
 	})
 
 	t.Run("WithTimeout", func(t *testing.T) {
-		t.Parallel()
+		ctx, cancel := clock.WithTimeout(context.Background(), 1*time.Millisecond)
+		defer cancel()
+
+		t.Run("Deadline", func(t *testing.T) {
+			dl, ok := ctx.Deadline()
+			assert.True(t, ok, "must have a deadline")
+			assert.True(t, dl.After(now), "deadline must be in the future")
+		})
 
-		ctx := context.Background()
-		ctx, cancel := clock.WithTimeout(ctx, time.Second)
+		advance(1 * time.Millisecond)
+
+		select {
+		case <-ctx.Done():
+			assert.Error(t, ctx.Err(), "done context must error")
+			assert.ErrorIs(t, ctx.Err(), context.DeadlineExceeded,
+				"context must have exceeded its deadline")
+
+		case <-time.After(10 * time.Millisecond):
+			t.Fatal("expected context to be done")
+		}
+	})
+
+	t.Run("WithTimeout/Value", func(t *testing.T) {
+		type contextKey string
+		key := contextKey("foo")
+
+		ctx1 := context.WithValue(context.Background(), key, "bar")
+
+		ctx2, cancel := clock.WithTimeout(ctx1, 1*time.Millisecond)
 		defer cancel()
 
-		_, ok := ctx.Deadline()
-		assert.True(t, ok, "must have deadline")
+		assert.Equal(t, "bar", ctx2.Value(key), "value must be preserved")
 	})
+
+	t.Run("WithTimeout/Cancel", func(t *testing.T) {
+		ctx, cancel := clock.WithTimeout(context.Background(), 1*time.Millisecond)
+		cancel()
+
+		select {
+		case <-ctx.Done():
+			assert.Error(t, ctx.Err(), "done context must error")
+			assert.ErrorIs(t, ctx.Err(), context.Canceled,
+				"context must have been canceled")
+
+		case <-time.After(10 * time.Millisecond):
+			t.Fatal("expected context to be done")
+		}
+	})
+}
+
+func TestMock_Sleep(t *testing.T) {
+	clock := NewMock()
+
+	ch := make(chan struct{})
+	go func() {
+		clock.Sleep(2 * time.Millisecond)
+		close(ch)
+	}()
+
+	// We cannot advance time until we're certain
+	// that the Sleep call has started waiting.
+	// Otherwise, we'll advance that one millisecond,
+	// and then the Sleep will start waiting for another Advance,
+	// which will never come.
+	//
+	// AwaitScheduled will block until there is at least one
+	// scheduled event.
+	clock.AwaitScheduled(1)
+
+	// Adbance only one millisecond, the Sleep should not return.
+	clock.Add(1 * time.Millisecond)
+	select {
+	case <-ch:
+		t.Fatal("sleep should not have returned")
+	case <-time.After(1 * time.Millisecond):
+		// ok
+	}
+
+	// Avance to the next millisecond, the Sleep should return.
+	clock.Add(1 * time.Millisecond)
+	select {
+	case <-ch:
+		// ok
+	case <-time.After(10 * time.Millisecond):
+		t.Fatal("expected Sleep to return")
+	}
+}
+
+func TestMock_AddNegative(t *testing.T) {
+	clock := NewMock()
+	assert.Panics(t, func() { clock.Add(-1) })
+}
+
+func TestMock_ManySleepers(t *testing.T) {
+	const N = 100
+
+	clock := NewMock()
+
+	var wg sync.WaitGroup
+	wg.Add(N)
+	for i := 0; i < N; i++ {
+		go func() {
+			defer wg.Done()
+
+			clock.Sleep(1 * time.Millisecond)
+		}()
+	}
+
+	clock.AwaitScheduled(N)
+	clock.Add(1 * time.Millisecond)
+	wg.Wait()
```

Comment:
Probably overkill, but should we instead use a channel and expect N messages? Then we can abort after a certain timeout and avoid deadlock if, for whatever reason, a regression causes Sleep to not return.
```
ch := make(chan struct{}, N)
for i := 0; i < N; i++ {
	go func() {
		clock.Sleep(1 * time.Millisecond)
		ch <- struct{}{}
	}()
}
clock.AwaitScheduled(N)
var got int
timeout := time.After(10 * time.Millisecond)
for got < N {
	select {
	case <-ch:
		got++
	case <-timeout:
		t.Fatal("expected all Sleeps to unblock")
	}
}
```

Reply:
A channel for that would work.

Although I think if the main concern is a deadlock, might make sense to stick with the wait group and just use a channel to enforce a wait specifically on that. I'll make that change.